### PR TITLE
Extract Initial Sync Into its Own Package

### PR DESCRIPTION
**Pull request by rawfalafel**
This addresses https://github.com/prysmaticlabs/prysm/issues/388

Once `InitialSyncService` is finished, it calls `Start` on `SyncService`. Note that `Start` for each sync service checks the state of the chain to decide whether or not to start the goroutine.

### CODE REVIEW

**Comment by rawfalafel** [comment]
This is ready for review. 

Currently, `Simulator` assumes an empty beacon chain and broadcasts blocks from the genesis block. However, `InitialSync` assumes that the network knows about a non-empty beacon chain where `slot_number > 0`. These two assumptions are incompatible at the moment. 

Although a new node connected to a network would run `InitialSync` to catch up to the network's longest chain before running `Sync`, our simulated node should skip `InitialSync` and immediately begin syncing from the genesis block. This PR includes those changes and some TODO statements for when we're ready to test our code with the assumption that the node is connected to a network.

File: beacon-chain/sync/initial-sync/service.go [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+	"fmt"
+	"time"
+
+	"github.com/prysmaticlabs/prysm/beacon-chain/types"
+	pb "github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1"
+	"github.com/prysmaticlabs/prysm/shared/p2p"
+	"github.com/sirupsen/logrus"
+)
+
+var log = logrus.WithField("prefix", "sync")
```
Comment (terenc3t) [inline beacon-chain/sync/initial-sync/service.go]: should this be "initialsync"?

`var log = logrus.WithField("prefix", "initialsync")`

File: beacon-chain/sync/initial-sync/service.go [hunk 2]
```diff
@@ -0,0 +1,265 @@
+package initialsync
```
Comment (terenc3t) [inline beacon-chain/sync/initial-sync/service.go]: can we add package level comment?

File: beacon-chain/sync/initial-sync/service.go [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+	}
+	log.WithField("Block received with hash", fmt.Sprintf("0x%x", h)).Debug("Crystallized state hash exists locally")
+
+	if err := s.writeBlockToDB(block); err != nil {
+		return err
+	}
+
+	s.initialCrystallizedStateHash = block.CrystallizedStateHash()
+
+	log.Infof("Saved block with hash 0%x for initial sync", h)
```
Comment (terenc3t) [inline beacon-chain/sync/initial-sync/service.go]: is 0 a typo? or should it be 0x?

File: beacon-chain/sync/initial-sync/service.go [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+
+	"github.com/prysmaticlabs/prysm/beacon-chain/types"
+	pb "github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1"
+	"github.com/prysmaticlabs/prysm/shared/p2p"
+	"github.com/sirupsen/logrus"
+)
+
+var log = logrus.WithField("prefix", "sync")
+
+// Config allows the channel's buffer sizes to be changed.
```
Comment (nisdas) [inline beacon-chain/sync/initial-sync/service.go]: Not sure what this config has to do with buffer sizes

File: beacon-chain/sync/initial-sync/service.go [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+// This method is normally called by the main node.
+func NewInitialSyncService(ctx context.Context,
+	cfg Config,
+	beaconp2p types.P2P,
+	chainService ChainService,
+	syncService SyncService,
+) *InitialSync {
+	ctx, cancel := context.WithCancel(ctx)
+
+	blockBuf := make(chan p2p.Message)
```
Comment (nisdas) [inline beacon-chain/sync/initial-sync/service.go]: shouldn't these channels be buffered ?

File: beacon-chain/sync/initial-sync/service.go [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+func (s *InitialSync) Start() {
+	stored, err := s.chainService.HasStoredState()
+	if err != nil {
+		log.Errorf("error retrieving stored state: %v", err)
+		return
+	}
+
+	if stored {
+		// TODO: Bail out of the sync service if the chain is only partially synced
+		log.Infof("chain state detected, exiting initial sync")
```
Comment (nisdas) [inline beacon-chain/sync/initial-sync/service.go]: First letter should be capitalized for logs

**Comment by rauljordan** [comment]
Hey @rawfalafel update on this PR?

**Comment by rawfalafel** [comment]
Going to address comments this afternoon!

2018年8月20日(月) 11:43 Raul Jordan <notifications@github.com>:

**Comment by rawfalafel** [comment]
ready for another look

**Pull request merged**